src/librustc/session/config.rs

```diff
@@ -1285,6 +1283,72 @@ pub fn parse_crate_types_from_list(list_list: Vec<String>) -> Result<Vec<CrateTy
     return Ok(crate_types);
 }
 
+pub mod nightly_options {
+    use getopts;
+    use syntax::feature_gate::UnstableFeatures;
+    use super::{ErrorOutputType, OptionStability, RustcOptGroup, get_unstable_features_setting};
+    use session::{early_error, early_warn};
+
+    pub fn is_unstable_enabled(matches: &getopts::Matches) -> bool {
+        is_nightly_build() && matches.opt_strs("Z").iter().any(|x| *x == "unstable-options")
+    }
+
+    fn is_nightly_build() -> bool {
+        match get_unstable_features_setting() {
+            UnstableFeatures::Allow | UnstableFeatures::Cheat => true,
+            _ => false,
+        }
+    }
+
+    fn check_enabled(matches: &getopts::Matches, enable_flag: &str, flag: &str) {
+        let e = if !is_nightly_build() {
+            format!("the option `{}` is only accepted on the nightly compiler", flag)
+        } else if !is_unstable_enabled(matches) {
+            format!("the `-Z {}` flag must also be passed to enable the flag `{}`",
+                    enable_flag, flag)
+        } else {
+            return
+        };
+        early_error(ErrorOutputType::default(), &e)
+    }
+
+    pub fn check_nightly_options(matches: &getopts::Matches, flags: &[RustcOptGroup]) {
+        for opt in flags.iter() {
+            if opt.stability == OptionStability::Stable {
+                continue
+            }
+            let opt_name = if opt.opt_group.long_name.is_empty() {
+                &opt.opt_group.short_name
+            } else {
+                &opt.opt_group.long_name
+            };
+            if !matches.opt_present(opt_name) {
+                continue
+            }
+            if opt_name != "Z" {
```

Comment:
This check continues to be different from where this was copy pasted from, the `check_enabled` function unconditionally produces an error on stable/beta builds, but the "unstable but not really" level of stability should only issue a warning on stable/beta.

I would encourage you to scrutinize the difference between this function and from where it came from, the behavior should continue to be the same after refactoring.